[Review] Remove java API use of deprecated RMM APIs [skip ci]

revans2 opened the pull request:
This removes all direct uses of the deprecated RMM APIs.  I cannot remove `-Wno-error=deprecated-declarations` yet as we include cudf headers that indirectly end up pulling in rmm.hpp by way of error.hpp which sill uses the deprecated APIs internally.

jlowe commented on java/src/main/java/ai/rapids/cudf/Rmm.java:
```diff
@@ -171,32 +235,76 @@ public static void shutdown(long forceGCInterval, long maxWaitTime, TimeUnit uni
+   * allocate something on is not always the one you want to synchronize against
+   * when you free it.
+   */
+  public static class RmmBuff implements AutoCloseable {
```
Should this be public? I'm thinking we only want to expose `DeviceMemoryBuffer`, or is there a use-case for apps to access `RmmBuff` directly that cannot be served via `DeviceMemoryBuffer`?  I wonder if there is a real need for `RmmBuff` or if these details should be moved into `DeviceMemoryBuffer` directly.

If this remains, I think we should spring for the two extra letters for consistency with `MemoryBuffer`, `DeviceMemoryBuffer`, and `HostMemoryBuffer`.

revans2 replied:
I made it public because the original RMM API was public.  I am happy to make it package private.

jlowe replied:
Is there a good reason for it to be separate from DeviceMemoryBuffer?  The more I think about it, the more it feels like RMM should just return that directly.

jlowe commented on java/src/main/java/ai/rapids/cudf/Rmm.java:
```diff
@@ -56,18 +95,62 @@ public static void initialize(int allocationMode, boolean enableLogging, long po
+    String path = null;
+    if (logConf != null) {
+      if (logConf.location != null) {
+        logTo = 1;
```
It would be cleaner and clearer if these magic values were a private enum (with their corresponding native values) and used directly in the `LogConf` class rather than separate double-booleans to track stdout vs. stderr (and the invalid state of stdout _and_ stderr).

jlowe commented on java/src/main/native/src/RmmJni.cpp:
```diff
@@ -266,55 +274,217 @@ void set_java_device_memory_resource(
+ virtual ~logging_resource_adaptor() {
+  if (close_at_end) {
+    delete out;
+    out = nullptr;
```
Should nulling the pointer be outside the `if`?

jlowe commented on java/src/main/native/src/RmmJni.cpp:
```diff
@@ -266,55 +274,217 @@ void set_java_device_memory_resource(
+};
+
+// Need to keep both separate so we can shut them down appropriately
+std::unique_ptr<logging_resource_adaptor> logging_memory_resource{};
```
Style nit: Elsewhere in this file we've been using a starting capital letter to indicate global variables and help reduce the chance of conflicting with local names.

revans2 replied:
You are going to have to remind me every time.  That is not a convention I have ever seen used before so I always forget.

mythrocks commented on java/src/main/native/src/RmmJni.cpp:
```diff
@@ -266,55 +274,217 @@ void set_java_device_memory_resource(
+
+ virtual ~logging_resource_adaptor() {
+  if (close_at_end) {
+    delete out;
```
Since we're (conditionally) deleting `out`, it would be safer to include/disallow a copy constructor and assignment operator, to protect against double deletes:
```cpp
private:
    logging_resource_adaptor(logging_resource_adaptor const&) = delete;
    logging_resource_adaptor& operator = (logging_resource_adaptor const&) = delete;
```

The generated defaults wouldn't be safe.

mythrocks replied:
Argh. We might also need to disable the move constructor, if we don't need it. Otherwise, the defaults will replicate `close_at_end`, and cause a double `delete`.

revans2 commented:
@jlowe @mythrocks I updated this to pull in https://github.com/rapidsai/cudf/pull/4815 in addition to addressing all of your comments.  Please take another look.  I marked this a WIP until its dependencies go in.

jlowe commented on java/src/main/java/ai/rapids/cudf/Rmm.java:
```diff
@@ -171,32 +250,79 @@ public static void shutdown(long forceGCInterval, long maxWaitTime, TimeUnit uni
+   * when you free it.
+   *
+   * This is not intended to be an API that others outside of CUDF consume.  It is
+   * not like to other more user facing APIs like MemoryBuffer.
```
"not like to other"

jlowe commented on java/src/main/native/src/RmmJni.cpp:
```diff
@@ -266,55 +364,232 @@ void set_java_device_memory_resource(
-    JNI_RMM_TRY(env, , rmmFinalize());
+    // Instead of trying to undo all of the adaptors that we added in reverse order
+    // we just reset the base adaptor so the others will not be called any more
+    // and then clean them up in really any order.  There shoudl be no interaction with
```
```suggestion
    // and then clean them up in really any order.  There should be no interaction with
```

jlowe commented on java/src/main/native/src/RmmJni.cpp:
```diff
@@ -266,55 +364,232 @@ void set_java_device_memory_resource(
+    // and then clean them up in really any order.  There shoudl be no interaction with
+    // RMM during this time anyways.
+    Initialized_resource.reset(new rmm::mr::cuda_memory_resource());
+    rmm::mr::set_default_resource(Initialized_resource.get());
```
Is it worth checking that the current default resource is the one we expect to be tearing down and throw/warn if it isn't?  Thinking of cases where the application code calls into separate native bits that interact with RMM and potentially install their own handler.

revans2 replied:
It does not fix any race it just moves it.  RMM Init is saying go stomp on top of what is there and replace it with what I want. How can I tell if what is there is I am supposed to be replacing?

jlowe replied:
I am thinking about this scenario:
# cudf inits RMM, installs handlers while remembering the original default handler
# app calls other native code outside of cudf that installs its own (potentially delegating?) default resource
# cudf shuts down RMM, installing the original default handler

At this point we've stomped on the app's default resource which could be bad, especially if the app's handler is going to delegate to our (now destroyed) resources and some resource allocated with the app resource is freed afterwards.  I was hoping to turn that undefined behavior into an assertion that when we go to shutdown, the current default resource is the one we last installed.  The java handler already does this check when it is deregistered, so I guess we're already covered there.

I admit this is a weird case and requires the app to reach around cudf to access RMM directly which seems unlikely.  Given it's already covered for the case when the Java handler is installed (which will be the common case), maybe nothing to do here.

revans2 replied:
I don't think this is a valid use case.  When someone asks us to initialize RMM or to shut it down through Java or otherwise that says to me that we are going to wipe out all of the resource_managers and put the default resource manager in a known state. That is what we do. RMM internally was doing the same thing, which is what I based this off of.

https://github.com/rapidsai/rmm/blob/63ebb53bf21a58b98b4596f7b49a46d1d821b05d/src/memory_manager.cpp#L135-L152

https://github.com/rapidsai/rmm/blob/63ebb53bf21a58b98b4596f7b49a46d1d821b05d/src/memory_manager.cpp#L158-L167

There is no thought for someone else may have poked at the stack.  If you are really scared that someone is going to do that, then yes we can try and guard against it, but I would argue that it is beyond the scope of this because we are on par with what existed previously.

jlowe commented on java/src/main/java/ai/rapids/cudf/NativeDepsLoader.java:
```diff
@@ -51,7 +51,6 @@ static synchronized void loadNativeDeps() {
           loadDep(os, arch, toLoad);
         }
         loaded = true;
-        Rmm.defaultInitialize();
```
2020 copyrights

revans2 commented on java/src/main/java/ai/rapids/cudf/DeviceMemoryBuffer.java:
```diff
@@ -30,19 +30,20 @@
 
   private static final class DeviceBufferCleaner extends MemoryBufferCleaner {
-    private long address;
+    private Rmm.RmmBuff buff;
```
I didn't think about it being a DeviceMemoryBuffer, but that is the ideal for it.  I had to make stream an actual CudaStream mostly so I could keep the different constructors to DeviceMemoryBuffer separate and clear.

revans2 commented:
@jlowe it turns out that it was caused by us mixing static linking of cudart and dynamic linking of it.  I made it so we will always be consistent.  That was not a fun bug to track down.

jlowe commented:
@kkraus14 this Java-only PR should be ready to go, and merging it will unblock Java cudf unit test failures triggered by recent RMM updates.  Thanks!

rongou commented:
@revans2 was there a reason to rewrite `logging_resource_adaptor` instead of using the one from RMM?

revans2 commented:
@rongou

> @revans2 was there a reason to rewrite `logging_resource_adaptor` instead of using the one from RMM?

Yes two reasons.  The first is that the header-only library, spdlog, is not included with rmm, so as soon as you try to even include the header file with `logging_resource_adaptor` in it you need to also pull in spdlog.  For rmm they included the library as part of the repo, but here in cudf it is not available, and it is not shipped with rmm so it was simpler to write my own than it was to try and pull in a new dependency.

The second is that `logging_resource_adaport` only supports writing to a file, and we needed to support stdout and stderr because it fits better with spark and how it does logging.  So if I wanted to use it I would also have to make changes to rmm to support stdout and stderr.  They would not be big changes, but all of it together was more then I wanted to worry about.

rongou commented:
Sounds like we should at least create an issue in RMM to merge the two?
